sw/device/tests/keymgr_sideload_otbn_test.c

```diff
@@ -100,11 +133,49 @@ static void test_otbn_with_sideloaded_key(dif_keymgr_t *keymgr,
 
   // Load the X25519 application.
   CHECK(otbn_load_app(otbn_ctx, kOtbnAppX25519) == kOtbnOk);
-
   // Run the OTBN app and retrieve the result.
   uint32_t result[8];
   CHECK(run_x25519_app(otbn_ctx, result) == true);
 
+  // Clear the sideload key and check that OTBN errors with the correct error
+  // code (`KEY_INVALID` bit 5 = 1).
+  CHECK_DIF_OK(
+      dif_keymgr_sideload_clear_set_enabled(keymgr, kDifToggleEnabled));
+  LOG_INFO("Clearing the Keymgr generated sideload keys.");
+  uint32_t at_clear_salt_result[8];
+  CHECK(run_x25519_app_clearkey(otbn_ctx, at_clear_salt_result) == false);
+  CHECK_ARRAYS_NE(result, at_clear_salt_result, ARRAYSIZE(result));
+
+  // Disable sideload key clearing
+  CHECK_DIF_OK(
+      dif_keymgr_sideload_clear_set_enabled(keymgr, kDifToggleDisabled));
+  LOG_INFO("Disable clearing the Keymgr generated sideload keys.");
+
+  // Clear the ERR bits REG and reset the OTBN interrupts (should not even be
+  // necessary for this recoverable software error)
+  mmio_region_write32(otbn_ctx->dif.base_addr, OTBN_ERR_BITS_REG_OFFSET, 0x0);
+  CHECK_DIF_OK(dif_otbn_reset(&otbn_ctx->dif));
+
+  // Regenerate the sideload key
+  keymgr_testutils_generate_versioned_key(keymgr, sideload_params);
+  LOG_INFO("Keymgr generated HW output for OTBN.");
+
+  // Confirm keymgr is in idle status and without error
+  dif_keymgr_status_codes_t keymgr_status;
+  CHECK_DIF_OK(dif_keymgr_get_status_codes(keymgr, &keymgr_status));
+  LOG_INFO("Keymgr Status: 0x%08x", keymgr_status);
+
+  // Confirm OTBN has transitioned to the idle state and is not locked
+  dif_otbn_status_t actual_otbn_status;
+  static const dif_otbn_status_t exp_otbn_status = kDifOtbnStatusIdle;
+  CHECK(dif_otbn_get_status(&otbn_ctx->dif, &actual_otbn_status) == kDifOk);
+  LOG_INFO("OTBN status: 0x%08x\n", actual_otbn_status);
+  CHECK(actual_otbn_status == exp_otbn_status);
+
+  uint32_t post_clear_salt_result[8];
+  CHECK(run_x25519_app(otbn_ctx, post_clear_salt_result) == true);
+  CHECK_ARRAYS_EQ(result, post_clear_salt_result, ARRAYSIZE(result));
```

Comment:
in your comment, do you mean this check fails?

Reply:
Yes, there are a lot of redundant checks in this code I put up because I was debugging - it can be cleaned up and refactored, it's still drafty. I tried to test generating a valid key after clear but OTBN still errors out with an invalid key error still. I don't see any DIFs that I could be missing out to do this before running the app on OTBN again. Which register actually is being read by the DIF - is it the `hw2reg.err_bits`? 

From the waveforms, it seems that the `err_bits_o.key_invalid` from the otbn_core never clears out because `err_bits_q` never clears out. `err_bits_q` would clear out if `start_i` would switch to '1' but also if `err_bits_d` would be all zero (otbn_core.sv lines 584-595) which is not the case. Despite `start_i` switching to '1' in the snapshot below for 1 cycle, `err_bits_q` does not clear because `err_bits_d` is still indicating an invalid key error and only switches to 0 after the `key_invalid `signal from the OTBN controller switches to '0' which is many clock cycles later. That leaves `err_bits_o.key_invalid` still stuck at '1', although the key is valid and `start_i` has switched to '1' when the OTBN is provided a valid key.  

I am probably missing something here, and I am also not familiar with the code and struggling to trace all the signals etc. - just trying to figure out how to get OTBN to use the new valid key. What am I missing to get this to work?

![image](https://user-images.githubusercontent.com/5173875/205158831-47504530-3498-4bab-a2ce-582dc4cfe333.png)

Reply:
@ctopal @GregAC I test clearing the sideload key for the OTBN here, then regenerating a new valid key and seeing that OTBN uses it. 

But OTBN error bits register still errors out with invalid key although a valid key is presented at the sideload port. I am chasing some of the waveforms, but it is not too easy to figure out why, I am not able to follow the error hierarchy easily. So far to me it seems that `otbn_core.err_bits_q` still retains the invalid key error bit (it doesn't clear at the pos edge of `start_i`) and so it gets read into the `hw2reg.err_bits` when `err_bits_ en` = '1'.

Reply:
To follow up, I modified the test to read the error register after disabling the clearing but I ignore it (still gives out an `0x20`) and continue to copy the output from the OTBN at the end, and it is actually as expected, and the test continues and passes to the end!

So it does seem like the error register `hw2reg.err_bits` indicates an invalid key error when  `done_core` = '1' at at the last instruction after OTBN completes its operation, when it shouldn't. Should we be handling the error bits register in a particular way in the DIFs/test to take care that this does not happen?

Reply:
thx. So OTBN does use a proper key again and results are correct, which is nice, but the invalid key error is still set, correct?

Reply:
maybe check the [keymgr_key_i](https://github.com/lowRISC/opentitan/blob/a6b908283fccba3f8b6b44052c6ad87276dc21e8/hw/ip/otbn/rtl/otbn.sv#L66) signal? the .valid should be 1 by then. if not, its an issue in the key manager creating this. 

This is where it starts to go down into generating the key invalid signal in otbn_controller: 
https://github.com/lowRISC/opentitan/blob/a6b908283fccba3f8b6b44052c6ad87276dc21e8/hw/ip/otbn/rtl/otbn.sv#L1129

Reply:
> thx. So OTBN does use a proper key again and results are correct, which is nice, but the invalid key error is still set, correct?

yes, exactly.

Reply:
> maybe check the [keymgr_key_i](https://github.com/lowRISC/opentitan/blob/a6b908283fccba3f8b6b44052c6ad87276dc21e8/hw/ip/otbn/rtl/otbn.sv#L66) signal? the .valid should be 1 by then. if not, its an issue in the key manager creating this.
> 
> This is where it starts to go down into generating the key invalid signal in otbn_controller:
> 
> https://github.com/lowRISC/opentitan/blob/a6b908283fccba3f8b6b44052c6ad87276dc21e8/hw/ip/otbn/rtl/otbn.sv#L1129

yes I have chased that. And as far as I can tell, the key generation from the key manager and relevant signals (to indicate that the key is valid) are as expected. I would narrow it to the OTBN controller.

Reply:
In this scenario, does OTBN still execute instructions while we are waiting for a new valid key? If so, I think it's sensible to have a `key_invalid` recoverable error, right?

Reply:
> In this scenario, does OTBN still execute instructions while we are waiting for a new valid key? If so, I think it's possible to have a `key_invalid` recoverable error, right?
> 
> Oh okay, I think I understand it better now. In this scenario we are running three back to back OTBN executions, with valid key, with invalid key and then with another valid key. Third one still shows a recoverable error even though we started OTBN with a valid key - possibly because it didn't clear out the error from the second run.
> 
> Am I getting the problem right?

Yes, exactly. The third operation still indicates the recoverable invalid key error even though it starts with a valid key generated from the key manager at the sideload port. OTBN uses it and computes correctly, but the error register at the end gets updated from an error signal `err_bits_q` (which still indicates the invalid key error, it doesn't clear out earlier despite all the other signals clearing) when `done_core` switches to '1'.

Reply:
Reviewing the RTL the behaviour it should exhibit is a little out of step with the documentation.

There's actually a couple of copies of the error bits. The OTBN internal copy and the software visible copy. When OTBN starts the internal copy is wiped, when it finished the internal copy is ORed with the software visible copy and the result written to the software visible copy. The software visible copy can be wiped by writing to the error bits register. Without the write you'd see this stuck bit behaviour. However that write is happening so something's not quite right. I'll do a run of this test and capture some waveforms and take a look.

Reply:
Thanks @GregAC, that will be very helpful! What I observe from the waveforms and RTL, is that the `start_i` for the core switches to 1 before that of the controller (intentionally as I see from the RTL, though not sure why)...so while `err_bits_q` of the core gets cleared, it gets the invalid key error bit from `err_bits_d` again in the next cycle (because these are fed from the controller and are not yet cleared). Then when `start_i` of the controller is 1, its `err_bits_q` get cleared and stay cleared in the next cycle because its `err_bits_d` are all-zero already. So  `err_bits_q` of the core gets stuck with that invalid key error bit, and revives it basically once `done_core` is asserted at the end of the OTBN operation. `err_bits_q` feeds the output error bits and the ERROR bits register, as far as I can tell.

Reply:
Yes this is a hardware issue, exactly as you describe @gdessouky, nice catch!. This has suffered from things being moved around during OTBN development. At one point in time `otbn_controller` was the main controller of OTBN in that it directly received the main start signal and issued done. It was also the only thing that generated errors.

As development continued `otbn_start_stop_control` was introduced to handle  things before and after actual execution (reseeding LFSRs, secure wipe etc) and `otbn_controller` was only used for actual execution stuff. So `otbn_controller` was given it's start signal by `otbn_start_stop_control` after the top-level start signal once pre execution tasks were complete.

We also started generating errors in various places and consolidated the bits at the `otbn_core` level hence the two sets of error bit flops.

Should there be an OTBN v2 or v1.5 it'd be nice to rationalise all of this but for now I think we just need a new error clear signal going into `otbn_controller`. I'll write up and issue and can do a PR for next week sometime, unless a fix is urgently needed.

Reply:
Thanks @GregAC for looking into it and confirming! I also now understand the rationale behind the 2 sets of error bit flops.

In the meantime, I will work around the error bits checking in the test and check it in with a TODO. Whenever you push a fix, I'll update the test.